subsys/bluetooth/controller/ll_sw/nordic/lll/lll_clock.c

```diff
@@ -15,25 +15,106 @@
 #include "common/log.h"
 #include "hal/debug.h"
 
-void lll_clock_wait(void)
+static void clock_ready(struct device *dev, clock_control_subsys_t subsys,
+			void *user_data);
+
+static struct device *dev;
+
+int lll_clock_init(void)
+{
+	int err;
+
+	dev = device_get_binding(DT_INST_0_NORDIC_NRF_CLOCK_LABEL);
+	if (!dev) {
+		return -ENODEV;
+	}
+
+	err = clock_control_on(dev, CLOCK_CONTROL_NRF_SUBSYS_LF);
+
+	return err;
+}
+
+int lll_clock_wait(void)
 {
 	static bool done;
 
 	if (done) {
-		return;
+		return 0;
 	}
 	done = true;
 
-	struct device *clock =
-			device_get_binding(DT_INST_0_NORDIC_NRF_CLOCK_LABEL);
+	struct k_sem sem_clock_wait;
+	struct clock_control_async_data async_data = {
+		.cb = clock_ready,
+		.user_data = &sem_clock_wait,
+	};
+	int err;
+
+	k_sem_init(&sem_clock_wait, 0, 1);
+
+	err = clock_control_async_on(dev, CLOCK_CONTROL_NRF_SUBSYS_LF,
+				     &async_data);
+	if (err) {
+		return err;
+	}
+
+	k_sem_take(&sem_clock_wait, K_FOREVER);
+
+	return 0;
+}
+
+int lll_hfclock_on(void)
+{
+	int err;
+
+	/* turn on radio clock in non-blocking mode. */
+	err = clock_control_on(dev, CLOCK_CONTROL_NRF_SUBSYS_HF);
+	if (!err || err == -EINPROGRESS) {
+		DEBUG_RADIO_XTAL(1);
+	}
+
+	return err;
+}
+
+int lll_hfclock_on_wait(void)
+{
+	struct k_sem sem_clock_wait;
```

Comment:
can you add assert when in irq context?

Reply:
Damn, I am in IRQ, I cant use semaphores to wait :-(

Reply:
@ioannisg I cant use k_sem_take inside ISR. Any alternative ideas to replace k_cpu_idle?

Reply:
isn't it used only in the test? Maybe you could handle it differently? `lll_hfclock_on(cb)`

Reply:
yes, will add it. The lll_clock_wait and lll_hfclock_on_wait shall only be called in thread.

Reply:
So: my "problem" with k_cpu_idle() is that it is "specified" to be used only in the Zephyr idle thread. Alternatively, you can _keep_ using the arch_cpu_idle _directly_ which is the ARM-implementation of k_cpu_idle, as long as you understand the consequences of it. Please, check https://github.com/zephyrproject-rtos/zephyr/pull/23511 for the details

Reply:
> Damn, I am in IRQ, I cant use semaphores to wait :-(

@cvinayak So the code is really supposed to busy-wait inside an ISR?

Reply:
@ioannisg no, it was my fault, only the call inside thread needs waiting. Calls from ISR do not need to wait, prepare will start and give enough time until event for the clock to settle. I will add assert checks and this PR (original commits) should hold good.

Reply:
@nordic-krch k_sem_take already has assert check on arch_is_in_isr. Hence, I dont need to add in controller.